### PR TITLE
[KMCompiler][Ascend][Nvidia]add unsqueeze_copy operator support

### PR DESCRIPTION
**Pull request by Keriza**
## PR Category

Operator

## Type of Change

New Feature

## Description

This PR introduces the `unsqueeze_copy` operator for NVIDIA CUDA and Ascend NPU backends, mirroring `torch.unsqueeze_copy`. It inserts a dimension of size 1 at the specified position and always returns a new tensor with independent storage (copy semantics) instead of a view. The implementation provides full PyTorch API compatibility:
- Correct dimension insertion
- Positive and negative dimension indexing
- Independent output storage
- Faithful copy semantics

### NVIDIA CUDA
- Triton kernel for explicit data copy
- Supports arbitrary tensor shapes and insertion dimensions
- Ensures output owns independent memory
- Supports `unsqueeze_copy.out`

### Ascend NPU
- Ascend backend with equivalent copy semantics
- Uses native NPU tensor operations
- Maintains PyTorch reference behaviour
- Output is allocated independently from the input

Supported data types: `float16`, `float32`, `bfloat16`

## Issue

N/A

## Progress

- [ ] Change is properly reviewed
- [ ] Change is responded to an issue
- [x] Change is fully covered by a UT

---

# Testing

Comprehensive validation performed on:
- NVIDIA CUDA
- Ascend 910B NPU

## Accuracy Test

```bash
pytest tests/test_unsqueeze_copy.py -v -s
```

### Validation Coverage

| Category            | Coverage                                          |
| :------------------ | :------------------------------------------------ |
| Backend             | NVIDIA CUDA, Ascend 910B NPU                      |
| Tensor shapes       | 1D, 2D, 3D, 4D tensors                           |
| Data types          | float16, float32, bfloat16                        |
| Dimension positions | Positive dimensions and negative dimension (`-1`) |

### Semantic Validation

- Output shape matches PyTorch reference
- Output device matches target backend
- `output.data_ptr() != input.data_ptr()`
- Modifying input does not affect output

### Results

- **NVIDIA CUDA: PASSED**
- **Ascend 910B NPU: PASSED**

---

# Performance

```bash
pytest benchmark/test_unsqueeze_copy.py -v -s
```

### Benchmark Coverage

| Category            | Coverage                                          |
| :------------------ | :------------------------------------------------ |
| Backend             | NVIDIA CUDA, Ascend 910B NPU                      |
| Data types          | float16, float32, bfloat16                        |
| Tensor shapes       | 1D, 2D, 3D, 4D tensors                           |
| Dimension positions | Positive dimensions and negative dimension (`-1`) |

The benchmark compares our implementation against the PyTorch reference operator `torch.ops.aten.unsqueeze_copy.default`, evaluating kernel latency, copy efficiency, and consistency across shapes and data types.

## NVIDIA CUDA Performance (H20 GPU)

The Triton kernel explicitly copies data into new storage while preserving copy semantics.

### float16

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.004896           | 0.005120          | 0.956        | [1024], dim=0        |
| SUCCESS | 0.004992           | 0.004992          | 1.000        | [1024], dim=-1       |
| SUCCESS | 0.006496           | 0.006752          | 0.962        | [1024,1024], dim=0   |
| SUCCESS | 0.006656           | 0.006528          | 1.020        | [1024,1024], dim=1   |
| SUCCESS | 0.006496           | 0.006560          | 0.990        | [1024,1024], dim=-1  |
| SUCCESS | 0.023456           | 0.022720          | 1.032        | [4096,4096], dim=0   |
| SUCCESS | 0.023584           | 0.022880          | 1.031        | [4096,4096], dim=-1  |
| SUCCESS | 0.040992           | 0.039616          | 1.035        | [32,1024,1024], dim=0 |
| SUCCESS | 0.041024           | 0.039712          | 1.033        | [32,1024,1024], dim=1 |
| SUCCESS | 0.040992           | 0.039712          | 1.032        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.040928           | 0.039680          | 1.031        | [1024,32,1024], dim=1 |
| SUCCESS | 0.010048           | 0.009952          | 1.010        | [8,32,128,128], dim=1 |
| SUCCESS | 0.010016           | 0.010048          | 0.997        | [8,32,128,128], dim=3 |
| SUCCESS | 0.009952           | 0.009856          | 1.010        | [8,32,128,128], dim=-1 |

### float32

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.005024           | 0.005216          | 0.963        | [1024], dim=0        |
| SUCCESS | 0.004896           | 0.005152          | 0.950        | [1024], dim=-1       |
| SUCCESS | 0.007968           | 0.007584          | 1.051        | [1024,1024], dim=0   |
| SUCCESS | 0.007872           | 0.007872          | 1.000        | [1024,1024], dim=1   |
| SUCCESS | 0.007808           | 0.007648          | 1.021        | [1024,1024], dim=-1  |
| SUCCESS | 0.041216           | 0.037808          | 1.090        | [4096,4096], dim=0   |
| SUCCESS | 0.041088           | 0.037312          | 1.101        | [4096,4096], dim=-1  |
| SUCCESS | 0.076256           | 0.068992          | 1.105        | [32,1024,1024], dim=0 |
| SUCCESS | 0.075904           | 0.069472          | 1.093        | [32,1024,1024], dim=1 |
| SUCCESS | 0.076224           | 0.068768          | 1.108        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.075904           | 0.069120          | 1.098        | [1024,32,1024], dim=1 |
| SUCCESS | 0.014560           | 0.013936          | 1.045        | [8,32,128,128], dim=1 |
| SUCCESS | 0.014496           | 0.014048          | 1.032        | [8,32,128,128], dim=3 |
| SUCCESS | 0.014400           | 0.013920          | 1.034        | [8,32,128,128], dim=-1 |

### bfloat16

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.005184           | 0.004992          | 1.038        | [1024], dim=0        |
| SUCCESS | 0.004992           | 0.004960          | 1.006        | [1024], dim=-1       |
| SUCCESS | 0.006720           | 0.006720          | 1.000        | [1024,1024], dim=0   |
| SUCCESS | 0.006560           | 0.006592          | 0.995        | [1024,1024], dim=1   |
| SUCCESS | 0.006560           | 0.006816          | 0.962        | [1024,1024], dim=-1  |
| SUCCESS | 0.023424           | 0.022784          | 1.028        | [4096,4096], dim=0   |
| SUCCESS | 0.023744           | 0.023072          | 1.029        | [4096,4096], dim=-1  |
| SUCCESS | 0.040928           | 0.039648          | 1.032        | [32,1024,1024], dim=0 |
| SUCCESS | 0.040960           | 0.039808          | 1.029        | [32,1024,1024], dim=1 |
| SUCCESS | 0.041088           | 0.039872          | 1.030        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.041120           | 0.039808          | 1.033        | [1024,32,1024], dim=1 |
| SUCCESS | 0.009984           | 0.009856          | 1.013        | [8,32,128,128], dim=1 |
| SUCCESS | 0.010176           | 0.009920          | 1.026        | [8,32,128,128], dim=3 |
| SUCCESS | 0.009952           | 0.010048          | 0.990        | [8,32,128,128], dim=-1 |

**Summary (NVIDIA CUDA):** The Triton implementation delivers latency comparable to or better than the PyTorch baseline across all tested shapes and data types, with improvements observed on selected larger tensor configurations.

## Ascend 910B NPU Performance

The Ascend implementation uses native NPU tensor operations, ensuring independent output storage and full `unsqueeze_copy` semantics.

### float16

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.042124           | 0.043707          | 0.964        | [1024], dim=0        |
| SUCCESS | 0.040895           | 0.042063          | 0.972        | [1024], dim=-1       |
| SUCCESS | 0.023085           | 0.022954          | 1.006        | [1024,1024], dim=0   |
| SUCCESS | 0.021455           | 0.022072          | 0.972        | [1024,1024], dim=1   |
| SUCCESS | 0.022707           | 0.023494          | 0.967        | [1024,1024], dim=-1  |
| SUCCESS | 0.023543           | 0.023694          | 0.994        | [4096,4096], dim=0   |
| SUCCESS | 0.023358           | 0.024090          | 0.970        | [4096,4096], dim=-1  |
| SUCCESS | 0.108143           | 0.108179          | 1.000        | [32,1024,1024], dim=0 |
| SUCCESS | 0.105587           | 0.105858          | 0.997        | [32,1024,1024], dim=1 |
| SUCCESS | 0.108075           | 0.108131          | 0.999        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.106489           | 0.106597          | 0.999        | [1024,32,1024], dim=1 |
| SUCCESS | 0.020245           | 0.020353          | 0.995        | [8,32,128,128], dim=1 |
| SUCCESS | 0.019047           | 0.019873          | 0.958        | [8,32,128,128], dim=3 |
| SUCCESS | 0.020108           | 0.020775          | 0.968        | [8,32,128,128], dim=-1 |

### float32

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.037819           | 0.038943          | 0.971        | [1024], dim=0        |
| SUCCESS | 0.037937           | 0.039100          | 0.970        | [1024], dim=-1       |
| SUCCESS | 0.021106           | 0.021748          | 0.970        | [1024,1024], dim=0   |
| SUCCESS | 0.020166           | 0.020250          | 0.996        | [1024,1024], dim=1   |
| SUCCESS | 0.021304           | 0.021630          | 0.985        | [1024,1024], dim=-1  |
| SUCCESS | 0.108176           | 0.108184          | 1.000        | [4096,4096], dim=0   |
| SUCCESS | 0.108416           | 0.108434          | 1.000        | [4096,4096], dim=-1  |
| SUCCESS | 0.214705           | 0.214678          | 1.000        | [32,1024,1024], dim=0 |
| SUCCESS | 0.214852           | 0.214839          | 1.000        | [32,1024,1024], dim=1 |
| SUCCESS | 0.214698           | 0.214673          | 1.000        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.214841           | 0.214817          | 1.000        | [1024,32,1024], dim=1 |
| SUCCESS | 0.021661           | 0.021442          | 1.010        | [8,32,128,128], dim=1 |
| SUCCESS | 0.021243           | 0.021237          | 1.000        | [8,32,128,128], dim=3 |
| SUCCESS | 0.019909           | 0.020120          | 0.989        | [8,32,128,128], dim=-1 |

### bfloat16

| Status  | Torch Latency (ms) | Gems Latency (ms) | Gems Speedup | Size Detail          |
| :------ | -----------------: | ----------------: | -----------: | :------------------- |
| SUCCESS | 0.039344           | 0.040399          | 0.974        | [1024], dim=0        |
| SUCCESS | 0.038171           | 0.038627          | 0.988        | [1024], dim=-1       |
| SUCCESS | 0.020175           | 0.021010          | 0.960        | [1024,1024], dim=0   |
| SUCCESS | 0.019732           | 0.020755          | 0.951        | [1024,1024], dim=1   |
| SUCCESS | 0.019880           | 0.020340          | 0.977        | [1024,1024], dim=-1  |
| SUCCESS | 0.019532           | 0.019684          | 0.992        | [4096,4096], dim=0   |
| SUCCESS | 0.018780           | 0.019991          | 0.939        | [4096,4096], dim=-1  |
| SUCCESS | 0.107660           | 0.107702          | 1.000        | [32,1024,1024], dim=0 |
| SUCCESS | 0.106213           | 0.106450          | 0.998        | [32,1024,1024], dim=1 |
| SUCCESS | 0.107830           | 0.107812          | 1.000        | [32,1024,1024], dim=-1 |
| SUCCESS | 0.106638           | 0.106825          | 0.998        | [1024,32,1024], dim=1 |
| SUCCESS | 0.020333           | 0.021066          | 0.965        | [8,32,128,128], dim=1 |
| SUCCESS | 0.020311           | 0.020167          | 1.007        | [8,32,128,128], dim=3 |
| SUCCESS | 0.019853           | 0.020497          | 0.969        | [8,32,128,128], dim=-1 |

**Summary (Ascend NPU):** Performance remains comparable with the PyTorch baseline across all tested shapes and data types while preserving independent storage semantics.

---

# Files Changed

- `src/flag_gems/ops/unsqueeze_copy.py` – CUDA implementation, Triton copy kernel, and `unsqueeze_copy_out`.
- `src/flag_gems/runtime/backend/_ascend/ops/unsqueeze_copy.py` – Ascend NPU backend with independent output storage.
- `src/flag_gems/__init__.py` – Registered `unsqueeze_copy` and `unsqueeze_copy.out`.
- `src/flag_gems/ops/__init__.py` – Exported the new operator.
- `src/flag_gems/runtime/backend/_ascend/ops/__init__.py` – Registered Ascend backend operator.
- `conf/operators.yaml` – Added operator metadata.
- `tests/test_unsqueeze_copy.py` – Accuracy tests, storage independence, and NPU semantic validation.
- `benchmark/test_unsqueeze_copy.py` – Comprehensive performance benchmarks.
```

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -8934,6 +8934,12 @@ ops:
       - Math
     stages:
       - alpha: '5.4'
+  - id: unsqueeze_copy
```
Comment (Caeruleann) [inline conf/operators.yaml]: The YAML file is incomplete. Please complete the description for this ID.
Also add `unsqueeze_copy_out` for this file.

File: src/flag_gems/ops/__init__.py [hunk 2]
```diff
@@ -1255,6 +1256,8 @@
     "uniform_",
     "unique_consecutive",
     "unique_dim",
+    "unsqueeze_copy",
```
Comment (Caeruleann) [inline src/flag_gems/ops/__init__.py:1479]: Please sort the list.

File: src/flag_gems/runtime/backend/_ascend/ops/__init__.py [hunk 3]
```diff
@@ -190,6 +191,8 @@
     "threshold",
     "threshold_backward",
     "triu",
+    "unsqueeze_copy",
```
Comment (Caeruleann) [inline src/flag_gems/runtime/backend/_ascend/ops/__init__.py:208]: Please sort the list.

File: tests/test_unsqueeze_copy.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        pytest.skip("NPU unavailable")
+
+    inp = torch.randn(
+        shape,
+        device="npu",
+        dtype=torch.float32,
+    )
+
+    with flag_gems.use_gems():
+        out = flag_gems.unsqueeze_copy(inp, dim)
```
Comment (Caeruleann) [inline tests/test_unsqueeze_copy.py]:
```suggestion
        out = torch.unsqueeze_copy(inp, dim)
```

File: tests/test_unsqueeze_copy.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    with flag_gems.use_gems():
+        res_out = flag_gems.unsqueeze_copy(res_inp, dim)
+
+    utils.gems_assert_close(res_out, ref_out, dtype)
+
+    # unsqueeze_copy must allocate new storage
+    assert res_out.data_ptr() != res_inp.data_ptr()
+
+
+@pytest.mark.parametrize(
```
Comment (Caeruleann) [inline tests/test_unsqueeze_copy.py:37]: Please add @pytest.mark.unsqueeze_copy

File: tests/test_unsqueeze_copy.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    # storage independence
+    assert out.data_ptr() != inp.data_ptr()
+
+    # verify copy semantics
+    out_clone = out.clone()
+
+    inp.fill_(0)
+
+    assert torch.equal(out, out_clone)
```
Comment (Caeruleann) [inline tests/test_unsqueeze_copy.py:91]: Please add unsquezze_copy_out test case.

File: src/flag_gems/ops/unsqueeze_copy.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import torch
+import triton
+import triton.language as tl
+
+
+logger = logging.getLogger(__name__)
+
+
+@triton.jit
+def _copy_kernel(
```
Comment (Caeruleann) [inline src/flag_gems/ops/unsqueeze_copy.py]: The Ascend implementation uses copy_ for unsqueeze_copy. Do we need a custom Triton kernel for the generic path? Since this is mainly a copy operation, could we use the same copy_ approach for consistency?

**Review by Caeruleann — COMMENTED** [review]
Please fix the failing checks and address the review comments.

**Comment by bin913** [comment]
@Keriza Please resolve conflict and fix failing checks

**Comment by Keriza** [comment]
> @Keriza Please resolve conflict and fix failing checks

Noted. I'll resolve the conflicts and fix CI failures a bit later.

File: conf/operators.yaml (hunk 1 above)
Comment (Keriza) [reply conf/operators.yaml]: Updated `conf/operators.yaml` with the complete description for `unsqueeze_copy` and added the `unsqueeze_copy_out` entry.

File: tests/test_unsqueeze_copy.py (hunk 5 above)
Comment (Keriza) [reply tests/test_unsqueeze_copy.py:37]: Added the `@pytest.mark.unsqueeze_copy` marker to the corresponding tests.

File: tests/test_unsqueeze_copy.py (hunk 6 above)
Comment (Keriza) [reply tests/test_unsqueeze_copy.py:91]: Added the unsqueeze_copy_out test case to cover the out variant.

File: src/flag_gems/ops/__init__.py (hunk 2 above)
Comment (Keriza) [reply src/flag_gems/ops/__init__.py:1479]: > Please sort the list.

I think the list is already sorted based on the current order. Could you please clarify if there is a different ordering you expect?

File: tests/test_unsqueeze_copy.py (hunk 4 above)
Comment (Keriza) [reply tests/test_unsqueeze_copy.py]: Updated to use `torch.unsqueeze_copy` inside `flag_gems.use_gems()`.

File: src/flag_gems/ops/unsqueeze_copy.py (hunk 7 above)
Comment (Keriza) [reply src/flag_gems/ops/unsqueeze_copy.py]: Updated to use `copy_` for the generic path. The performance remains stable around 0.9x-1.0x.
